doc/services/smf/index.rst

```diff
@@ -124,13 +131,41 @@ State Machine Termination
 =========================
 
 To terminate the state machine, the :c:func:`smf_set_terminate` function
-should be called. It can be called from the entry, run, or exit action. The
-function takes a non-zero user defined value that's returned by the
-:c:func:`smf_run_state` function. The function has the following prototype:
-``void smf_set_terminate(smf_ctx *ctx, int32_t val)``
+should be called. It can be called from the entry, run, or exit actions. The
+function takes a non-zero user defined value that will be returned by the
+:c:func:`smf_run_state` function.
+
+UML State Machines
+==================
+
+SMF follows UML hierarchical state machine rules for transitions i.e., the
+entry and exit actions of the least common ancestor are not executed on
+transition, unless said transition is a transition to self.
+
+The UML Specification for StateMachines may be found in chapter 14 of the UML
+specification available here: https://www.omg.org/spec/UML/
+
+SMF breaks from UML rules in:
+
+1. Executing the actions associated with the transition within the context
+   of the source state, rather than after the exit actions are performed.
+2. Only allowing external transitions to self, not to sub-states. A transition
+   from a superstate to a child state is treated as a local transition.
+3. Requiring a containing superstate for all active states (to provide a Least
+   Common Ancestor for all possible transitions).
```

Comment:
This text to me implies that the SMF requires a superstate for all active states while UML does not.  But the implementation of `get_lca_of()` can return NULL and `SMF_CREATE_STATE` allows setting the parent to NULL.

Reply:
Good point. I'll review

Reply:
The logic gets slightly convoluted if we don't force an HSM to have an enclosing root state.

You have to find the parent state of the source and destination states:
1. If they are the same, find the LCA and perform entry and exit actions up to but not including the LCA
2. If they are different perform all exit actions up to the parent state of source, then all entry actions from parent state of destination down to the destination.

It's probably not too hard, but slows the state machine operation down as you have to iterate through the states a few more times.

It may make sense to refactor the whole 'search for LCA' algorithm. I'm not too keen on doing that, but I'm not too keen on imposing the root state requirement either.

This feels like it would be easier if I could store an array of parent states for the transition source, to check the transition destination parents against, but I know RAM is at a premium.

Opinions welcome 🙂